### ISSUE TITLE
Can I use STL file directly to analyze radiomics?

### ISSUE THREAD

**Issue by sinsuz2021** [SimpleITK/SimpleITK#1439]
I want to obtain feature of radiomics.

I found an example of code for performing on Google colaboratory in a web page:

========================================
========================================

!pip install pyradiomics==2.1.0

import os  # needed navigate the system to get the input data
from radiomics import featureextractor  # This module is used for interaction with pyradiomics
import radiomics
from radiomics import firstorder, getTestCase, glcm, glrlm, glszm, imageoperations, shape, shape2D
import numpy as np
import six

#http://simpleitk.github.io/SimpleITK-Notebooks/01_Image_Basics.html
import SimpleITK as sitk
import matplotlib.pyplot as plt

# brain1, brain2, breast1, lung1,lung2,(prostate_phantom)
imageName, maskName = getTestCase('lung2')
image = sitk.ReadImage(imageName)
mask = sitk.ReadImage(maskName)

ndImg = sitk.GetArrayFromImage(image)
ndLbl = sitk.GetArrayFromImage(mask)
plt.imshow(ndImg[24])
plt.show()
plt.imshow(ndLbl[24])
plt.show()

settings = {}
settings['binWidth'] = 25
# If enabled, resample image (resampled image is automatically cropped.
settings['resampledPixelSpacing'] = None  # [3,3,3] is an example for defining resampling (voxels with size 3x3x3mm)
settings['interpolator'] = sitk.sitkBSpline
settings['label'] = 1 

bb, correctedMask = imageoperations.checkMask(image, mask)
if correctedMask is not None:
  mask = correctedMask
image, mask = imageoperations.cropToTumorMask(image, mask, bb)

firstOrderFeatures = firstorder.RadiomicsFirstOrder(image, mask, **settings)
# firstOrderFeatures.enableFeatureByName('Mean', True)
firstOrderFeatures.enableAllFeatures()

print('Will calculate the following first order features: ')
for f in firstOrderFeatures.enabledFeatures.keys():
  print('  ', f)
  print(getattr(firstOrderFeatures, 'get%sFeatureValue' % f).__doc__)

print('Calculating first order features...')
results = firstOrderFeatures.execute()
print('done')

print('Calculated first order features: ')
for (key, val) in six.iteritems(results):
  print('  ', key, ':', val)

========================================
========================================

This sample code worked well.

But I have own data in which the object (to be analyzed) was already extracted as STL file.
![stl_file](https://user-images.githubusercontent.com/88314291/128986535-b4ddce6f-0360-4b86-b0db-558a4723f7eb.png)

This image was displayed on my space of Google Colaboratory by following code:

========================================
========================================

!pip install pyradiomics
import os
from radiomics import featureextractor
import radiomics
from radiomics import firstorder, getTestCase, glcm, glrlm, glszm, imageoperations, shape, shape2D

import numpy as np
import six
import SimpleITK as sitk
import matplotlib.pyplot as plt

!pip install  numpy-stl

from stl import mesh
from mpl_toolkits import mplot3d

figure = pyplot.figure()
axes = mplot3d.Axes3D(figure)

your_mesh = mesh.Mesh.from_file('/content/drive/MyDrive/Colab Notebooks/20210811PVICT_STL/20191002-66694.stl') 

def mesh_location_zero(your_mesh):
    midPosRel = (your_mesh.max_ - your_mesh.min_)/2
    your_mesh.x = your_mesh.x - (midPosRel[0] + your_mesh.min_[0])
    your_mesh.y = your_mesh.y - (midPosRel[1] + your_mesh.min_[1])
    your_mesh.z = your_mesh.z - (midPosRel[2] + your_mesh.min_[2])
    return your_mesh

mesh_location_zero(your_mesh)
axes.add_collection3d(mplot3d.art3d.Poly3DCollection(your_mesh.vectors))

scale = your_mesh.points.flatten()
axes.auto_scale_xyz(scale, scale, scale)

pyplot.show()

========================================
========================================

Can I use this file to obtain the feature of radiomics?

I intended to use followings... 
========================================
imagePath = "/content/drive/MyDrive/Colab Notebooks/20210811PVICT_STL/20191002-66694.stl"
maskPath =  "/content/drive/MyDrive/Colab Notebooks/20210811PVICT_STL/20191002-66694copy.stl"
image = sitk.ReadImage(imagePath)
mask = sitk.ReadImage(maskPath)
========================================

...instead of followings.
========================================
imageName, maskName = getTestCase('lung2')
image = sitk.ReadImage(imageName)
mask = sitk.ReadImage(maskName)
========================================

But, when I run the code of "image = sitk.ReadImage(imagePath)",
runtime error below is shown:
========================================
RuntimeError: Exception thrown in SimpleITK ImageFileReader_Execute: /tmp/SimpleITK/Code/IO/src/sitkImageReaderBase.cxx:105:
sitk::ERROR: Unable to determine ImageIO reader for "/content/drive/MyDrive/Colab Notebooks/20210811PVICT_STL/20191002-66694.stl"
========================================

**Comment by dave3d** [comment]
Another approach is to use VTK to convert a 3d mesh to a 3d image.  You would use VTK's PolyDataToImageDataStencil class.  Here is an example that uses the class:

https://kitware.github.io/vtk-examples/site/Python/PolyData/PolyDataToImageDataStencil/

**Comment by sinsuz2021** [comment]
Hello, zivy
Hello, dave3d
Thank you so much for your early reply!
At first, I will try both of your methods.
I will report the results later.

**Comment by sinsuz2021** [comment]
Hello, zivy
Thank you so much for your advice!
I am sorry to be late for replying...
I tried the method you suggested.

Using 3D Slicer, DICOM file was easily loaded, and STL file was also easily loaded (converted from "model" to segmentation in "Data" interface of 3D Slicer).
Finally, I could obtain two files: 
20191002-66694.nrrd (image file; file name was changed to suit for coding)
20191002-66694-segmentation.seg.nrrd (label file)

I run the code below:
========================================
imagePath = "/content/drive/MyDrive/Colab Notebooks/20210818_20191002-66694/20191002-66694.nrrd"
maskPath = "/content/drive/MyDrive/Colab Notebooks/20210818_20191002-66694/20191002-66694-segmentation.seg.nrrd"
image = sitk.ReadImage(imagePath)
mask = sitk.ReadImage(maskPath)
========================================
It did work!!!
Thank you very much!!!


But when I run the code below...
========================================
ndImg = sitk.GetArrayFromImage(image)
ndLbl = sitk.GetArrayFromImage(mask)
plt.imshow(ndImg[24])
plt.show()
plt.imshow(ndLbl[24])
plt.show()
========================================

The size of the image and label was different!
![image_and_label](https://user-images.githubusercontent.com/88314291/129885885-84422838-2b49-4f48-a431-203e17bcb653.PNG)


Further, I run the code:
========================================
settings = {}
settings['binWidth'] = 25
settings['resampledPixelSpacing'] = None
settings['interpolator'] = sitk.sitkBSpline
settings['label'] = 1 
========================================
It was done without error.

But, when I run the code below, error message appeared.
========================================
bb, correctedMask = imageoperations.checkMask(image, mask)
if correctedMask is not None:
  mask = correctedMask
image, mask = imageoperations.cropToTumorMask(image, mask, bb)
========================================

The error message was below:
========================================
RuntimeError: Exception thrown in SimpleITK LabelStatisticsImageFilter_Execute: /tmp/SimpleITK/Code/BasicFilters/src/sitkImageFilter.cxx:63:
sitk::ERROR: Input "labelImage" for "LabelStatisticsImageFilter" has size of [ 339, 198, 231 ] which does not match the primary input's size of [ 512, 512, 320 ]!
========================================

It seems to be derived from the different size between image and label.

I am happy if you have a good advise where I can adjust the size of the label.
Of note, in the 3D slicer, the segmentation was just fitted to the image!

**Comment by zivy** [comment]
Hello @sinsuz2021,

In the future, please use [GitHub markdown for code](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) so that the messages are more readable.

You need to resample the segmentation onto the image using the identity transformation and then visualize:
```
imagePath = "/content/drive/MyDrive/Colab Notebooks/20210818_20191002-66694/20191002-66694.nrrd"
maskPath = "/content/drive/MyDrive/Colab Notebooks/20210818_20191002-66694/20191002-66694-segmentation.seg.nrrd"
image = sitk.ReadImage(imagePath)
mask = sitk.ReadImage(maskPath)
resampled_mask = sitk.Resample(mask, image, sitk.Transform(), sitk.sitkNearestNeighbor)
```
For various segmentation visualization options, please see [this Jupyter notebook](https://github.com/InsightSoftwareConsortium/SimpleITK-Notebooks/blob/master/Python/05_Results_Visualization.ipynb).

**Comment by dave3d** [comment]
Note that when converting from SimpleITK's image to a NumPy array, any meta-data about the image is lost.  In particular, the spatial information such as pixel spacing is gone.  So in SimpleITK the images might align perfectly finely, but when convert to NumPy, they will not.

**Comment by sinsuz2021** [comment]
Hello @zivy
Hello @dave3d

Thank you so much for your kind and valuable help!!!
By using the resampling code by @zivy, I could successfully obtain the radiomics features.
And by the helpful comments by @dave3d, I can understand the importance of resampling.

In the original code I referred to, I can find the code of resampling, which I ignored and I could not arrange for my circumstance.
Now I understand its importance and @zivy kindly suggested an alternative way!

Since I have to convert 500 sets of DICOM and STL files, and obtain radiomics features from them, 
I have to go on.
If I have a chance, I will try another way suggested by @dave3d and various segmentation visualization options suggested by @zivy!

Thank you very much!!!
